SpatialGDK/Source/Private/Interop/SpatialStaticComponentView.cpp

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "SpatialStaticComponentView.h"
+
+
+Worker_Authority USpatialStaticComponentView::GetAuthority(Worker_EntityId EntityId, Worker_ComponentId ComponentId)
+{
+	if (TMap<Worker_ComponentId, Worker_Authority>* ComponentAuthorityMap = EntityComponentAuthorityMap.Find(EntityId))
+	{
+		if (Worker_Authority* Authority = ComponentAuthorityMap->Find(ComponentId))
+		{
+			return *Authority;
+		}
+	}
+
+	return WORKER_AUTHORITY_NOT_AUTHORITATIVE;
+}
+
+// TODO UNR-640 - Need to fix for authority loss imminent
+bool USpatialStaticComponentView::HasAuthority(Worker_EntityId EntityId, Worker_ComponentId ComponentId)
+{
+	return GetAuthority(EntityId, ComponentId) == WORKER_AUTHORITY_AUTHORITATIVE;
+}
+
+template <typename T>
+typename T* USpatialStaticComponentView::GetComponentData(Worker_EntityId EntityId)
+{
+	if (TMap<Worker_ComponentId, TUniquePtr<ComponentStorageBase>>* ComponentStorageMap = EntityComponentMap.Find(EntityId))
+	{
+		if (TUniquePtr<improbable::ComponentStorageBase>* Component = ComponentStorageMap->Find(T::ComponentId))
+		{
+			return &(static_cast<improbable::ComponentStorage<T>*>(Component->Get())->Get());
+		}
+	}
+
+	return nullptr;
+}
+
+void USpatialStaticComponentView::OnAddComponent(const Worker_AddComponentOp& Op)
+{
+	TUniquePtr<improbable::ComponentStorageBase> Data;
+	switch (Op.data.component_id)
+	{
+	case SpatialConstants::ENTITY_ACL_COMPONENT_ID:
+		Data = MakeUnique<improbable::ComponentStorage<improbable::EntityAcl>>(Op.data);
+		break;
+	case SpatialConstants::METADATA_COMPONENT_ID:
+		Data = MakeUnique<improbable::ComponentStorage<improbable::Metadata>>(Op.data);
+		break;
+	case SpatialConstants::POSITION_COMPONENT_ID:
+		Data = MakeUnique<improbable::ComponentStorage<improbable::Position>>(Op.data);
+		break;
+	case SpatialConstants::PERSISTENCE_COMPONENT_ID:
+		Data = MakeUnique<improbable::ComponentStorage<improbable::Persistence>>(Op.data);
+		break;
+	case SpatialConstants::ROTATION_COMPONENT_ID:
+		Data = MakeUnique<improbable::ComponentStorage<improbable::Rotation>>(Op.data);
+		break;
+	case SpatialConstants::UNREAL_METADATA_COMPONENT_ID:
+		Data = MakeUnique<improbable::ComponentStorage<improbable::UnrealMetadata>>(Op.data);
+		break;
+	default:
+		return;
+	}
+
+	EntityComponentMap.FindOrAdd(Op.entity_id).FindOrAdd(Op.data.component_id) = std::move(Data);
+}
+
+void USpatialStaticComponentView::OnRemoveEntity(const Worker_RemoveEntityOp& Op)
+{
+	EntityComponentMap.Remove(Op.entity_id);
+}
+
+void USpatialStaticComponentView::OnComponentUpdate(const Worker_ComponentUpdateOp& Op)
+{
+	switch (Op.update.component_id)
+	{
+	case SpatialConstants::ENTITY_ACL_COMPONENT_ID:
+		if (improbable::EntityAcl* aclComponent = GetComponentData<improbable::EntityAcl>(Op.entity_id))
+		{
+			aclComponent->ApplyComponentUpdate(Op.update);
```

Comment:
What if we just grabbed the base component data, and add a virtual ApplyComponentUpdate function to the component class?

Reply:
Yeah, I planned to do this for a load of the repeated component methods but didn't get around to it. Will at least do if for `ApplyComponentUpdate` before merging.